### PR TITLE
feat: add bundle `--component-names-strategy` flag

### PR DESCRIPTION
**Pull request by kanoru3101**
## What/Why/How?
Added an opt-in `--component-names-strategy` flag to `bundle`: inlined Schema components are named from each schema's `title` (PascalCased format) instead of the file basename.
## Reference
#2797

### With flag

Screenshot 2026-06-03 at 22 07 20

### Error `missing title`

Screenshot 2026-06-03 at 23 32 15

### Warn `title collision`

Screenshot 2026-06-10 at 11 01 39

### Warn `two title collision`

Screenshot 2026-06-10 at 11 02 47

## Check yourself

- [x] This PR follows the [contributing guide](https://github.com/Redocly/redocly-cli/blob/main/CONTRIBUTING.md#pull-request-guidelines) 
- [x] All new/updated code is covered by tests
- [x] Core code changed? - Tested with other Redocly products (internal contributions only)
- [ ] New package installed? - Tested in different environments (browser/node)
- [x] Documentation update has been considered 

## Security

- [x] The security impact of the change has been considered 
- [x] Code follows company security practices and guidelines

---

> [!NOTE]
> **Low Risk**
> Opt-in bundle flag with default unchanged; scope is schema naming during bundling with tests and docs, not auth or runtime API behavior.
> 
> **Overview**
> Adds **`--component-names-strategy`** to `redocly bundle` (default **`basename`**, opt-in **`title`**) so inlined **Schema** components can be named from each schema’s **`title`** (PascalCased and sanitized for valid component keys) instead of file basename / pointer.
> 
> The bundler gains title-based naming with errors for missing `title`, warnings (or errors via existing **`--component-renaming-conflicts-severity`**) for title collisions, and basename fallback when title naming fails; **parameters** and other component types keep the previous behavior. Shared **`COMPONENT_NAME_CHARS`** is used for sanitization and the **`spec-components-invalid-map-name`** rule.
> 
> CLI wiring, bundle docs, changeset, unit/e2e tests, and help snapshots are updated.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9612a51ac0727020f3c7231216d4d4e383a4c960. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: docs/@v2/commands/bundle.md [hunk 1]
```diff
@@ -147,3 +148,31 @@ You can adjust how the CLI handles these naming conflicts with the `--component-
...
+# schemas/requests/Authority.yaml
+type: object
+title: Authority request
+```
+
+```bash
+redocly bundle openapi.yaml -o bundled.yaml --use-titles-for-component-names
+```
+
+The output now uses `AuthorityModel` and `AuthorityRequest`. A `title` containing `.`, `-`, or `_` keeps them — `order-item` becomes `Order-item`.
```
Comment (JLekawa) [inline docs/@v2/commands/bundle.md]:
```suggestion
The output uses `AuthorityModel` and `AuthorityRequest`.
A `title` containing `.`, `-`, or `_` keeps them — `order-item` becomes `Order-item`.
```

File: docs/@v2/commands/bundle.md [hunk 2]
```diff
@@ -147,3 +148,31 @@ You can adjust how the CLI handles these naming conflicts with the `--component-
 - `off`: No warnings or errors are shown.
 - `warn` (default): Shows a warning and renames conflicting components automatically.
 - `error`: Treats conflicts as errors; the bundling process fails if a naming conflict is detected.
+
+### Use titles for component names
+
+By default, the bundler names each inlined Schema component after its `$ref` — the JSON Pointer fragment, or the file's basename when there's no fragment. Two files that share a basename — for example `schemas/models/Authority.yaml` and `schemas/requests/Authority.yaml` — collide into `Authority` and `Authority-2`. That auto-numbered suffix is brittle: an unrelated `$ref` change can renumber it.
```
Comment (JLekawa) [inline docs/@v2/commands/bundle.md]:
```suggestion
By default, the bundler names each inlined Schema component after its `$ref` — the JSON Pointer fragment, or the file's basename when there's no fragment.
Two files that share a basename — for example `schemas/models/Authority.yaml` and `schemas/requests/Authority.yaml` — collide into `Authority` and `Authority-2`.
The auto-numbered suffix is brittle: an unrelated `$ref` change can renumber it.
```

**Review by JLekawa — APPROVED** [review]

**Comment by tatomyr** [comment]
@kanoru3101 let's check bugbot's comments. Also, please do not just resolve them, but leave meaningful comments.

File: packages/core/src/bundle/bundle-visitor.ts [hunk 3]
```diff
@@ -297,9 +305,30 @@ export function makeBundleVisitor({
...
+        const title = (target.node as { title?: string }).title;
+        ctx.report({
+          message:
+            `Title "${title}" maps to component name \`${titleName}\`, ` +
+            `already used by another schema. Rename one of the titles.`,
+          location: target.location.child('title'),
+          from: titleNameLocations.get(titleName),
+          forceSeverity: 'error',
+        });
+        return target.location.absolutePointer;
```
[the remark this answers is not in the document]
Comment (kanoru3101) [reply packages/core/src/bundle/bundle-visitor.ts]: It's already fixed.
If user uses `--force` with `--use-titles-for-component-names`, it creates component names with a suffix.

File: docs/@v2/commands/bundle.md [hunk 4]
```diff
@@ -147,3 +148,34 @@ You can adjust how the CLI handles these naming conflicts with the `--component-
...
+redocly bundle openapi.yaml -o bundled.yaml --use-titles-for-component-names
+```
+
+The output uses `AuthorityModel` and `AuthorityRequest`.
+A `title` containing `.`, `-`, or `_` keeps them — `order-item` becomes `Order-item`.
+
+Bundling fails (no output file unless you pass `--force`) when:
+
+- A referenced schema has no `title`.
+- A `title` uses anything other than ASCII letters, digits, spaces, `.`, `-`, or `_` (for example `&` or a non-ASCII title).
```
Comment (tatomyr) [inline docs/@v2/commands/bundle.md]: Could you explain your consideration for choosing this characters set?

File: packages/cli/src/commands/bundle.ts [hunk 5]
```diff
@@ -39,6 +39,7 @@ export type BundleArgv = {
   'remove-unused-components'?: boolean;
   'keep-url-references'?: boolean;
   'component-renaming-conflicts-severity'?: RuleSeverity;
+  'use-titles-for-component-names'?: boolean;
```
Comment (tatomyr) [inline packages/cli/src/commands/bundle.ts]: Could this option conflict with other options?

File: packages/core/src/bundle/bundle-title-naming.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import type { Location } from '../ref-utils.js';
+import type { UserContext } from '../walk.js';
+
+/** Build a PascalCase component key from a `title`; `null` if empty or has unsupported characters. */
```
Comment (tatomyr) [inline packages/core/src/bundle/bundle-title-naming.ts]: - Use a meaningful name so we don't need comments. 
- Consider making this function general (not only for titles but for any kind of strings), maybe it can be reused.
- Null is not a valid object key, isn't any better option out there?

**Review by tatomyr — COMMENTED** [review]
Could there be fewer changes? It's only adding a new option, so I'd suggest isolating everything related to it so the changes are not scattered too much. 
Also, please make sure all tests are meaningful and don't duplicate each other.

File: docs/@v2/commands/bundle.md (hunk 4 above)
Comment (kanoru3101) [reply docs/@v2/commands/bundle.md]: The character it's the Components Object key pattern from the `OpenAPI` and `AsyncAPI` specs `(^[a-zA-Z0-9.\-_]+$)`, the same one the `spec-components-invalid-map-name` rule uses.

File: packages/cli/src/commands/bundle.ts (hunk 5 above)
Comment (kanoru3101) [reply packages/cli/src/commands/bundle.ts]: Tested with different flags, seems to work without conflicts

File: packages/core/src/bundle/bundle-visitor.ts [hunk 7]
```diff
@@ -297,9 +305,32 @@ export function makeBundleVisitor({
...
+        const title = (target.node as { title?: string }).title;
+        ctx.report({
+          message:
+            `Title "${title}" maps to component name \`${titleName}\`, ` +
+            `already used by another schema. Rename one of the titles.`,
+          location: target.location.child('title'),
+          from: titleNameLocations.get(titleName),
+          forceSeverity: 'error',
+        });
+        return baseName;
```
[the remark this answers is not in the document]
Comment (kanoru3101) [reply packages/core/src/bundle/bundle-visitor.ts]: added fallback

File: packages/core/src/bundle/bundle-visitor.ts [hunk 8]
```diff
@@ -291,22 +299,84 @@ export function makeBundleVisitor({
...
+    target: ComponentTarget,
+    componentsGroup: Record<string, unknown>,
+    ctx: UserContext,
+    problem: { message: string; location: Location; from?: Location }
+  ) {
+    const name = componentNameFromBasename(target, componentsGroup, ctx).name;
+    if (!componentsGroup[name]) {
+      ctx.report({ ...problem, forceSeverity: 'error' });
+      titleNameLocations.set(name, target.location);
+    }
```
[the remark this answers is not in the document]
Comment (kanoru3101) [reply packages/core/src/bundle/bundle-visitor.ts:349]: The error is reported per output component, not per source file. The bundler merges schemas into one component `a/Thing.yaml` and `b/Thing.yaml` both become a single `#/components/schemas/Thing`. So two identical sources are one component and get one error. Schemas with different content become different components and are each reported.

**Comment by kanoru3101** [comment]
> Could there be fewer changes? It's only adding a new option, so I'd suggest isolating everything related to it so the changes are not scattered too much. Also, please make sure all tests are meaningful and don't duplicate each other.

I rewrote the tests and shortened them. Some of the tests check error handling, account for conflicts, and cover various types of errors

File: packages/core/src/__tests__/fixtures/refs/title-naming-collision/schemas/a/User.yaml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+type: object
+title: User
+properties:
+  id:
```
Comment (tatomyr) [inline packages/core/src/__tests__/fixtures/refs/title-naming-collision/schemas/a/User.yaml]: Let's tie the property to the file name.

File: packages/core/src/__tests__/bundle.test.ts [hunk 10]
```diff
@@ -844,3 +844,73 @@ describe('sibling $ref resolution by spec', () => {
...
+    expect(Object.keys(parsed.components.schemas).sort()).toEqual([
+      'AuthorityModel',
+      'AuthorityRequest',
+    ]);
+    expect(
+      parsed.paths['/authorities'].get.responses['200'].content['application/json'].schema
+    ).toEqual({ $ref: '#/components/schemas/AuthorityModel' });
+    expect(
+      parsed.paths['/authorities'].post.requestBody.content['application/json'].schema
+    ).toEqual({ $ref: '#/components/schemas/AuthorityRequest' });
```
Comment (tatomyr) [inline packages/core/src/__tests__/bundle.test.ts]: Could you use inline snapshots instead?

File: packages/core/src/__tests__/utils.test.ts [hunk 11]
```diff
@@ -139,6 +140,20 @@ describe('utils', () => {
...
+  describe('toPascalCase', () => {
+    it.each([
+      ['Authority model', 'AuthorityModel'],
+      ['authority-model', 'Authority-model'],
+      ['user_profile', 'User_profile'],
+      ['foo.bar', 'Foo.bar'],
+      ['API v2 User', 'APIV2User'],
+      ['  padded  ', 'Padded'],
+      ['', ''],
+    ])('converts %j to %j', (input, expected) => {
```
Comment (tatomyr) [inline packages/core/src/__tests__/utils.test.ts]: What is `%j`?

File: packages/core/src/bundle/bundle-visitor.ts [hunk 12]
```diff
@@ -206,7 +223,7 @@ export function makeBundleVisitor({
   };
 
   if (version === 'oas3') {
-    const componentType = mapTypeToComponent('Schema', version)!;
+    const componentType = schemaComponentType!;
```
Comment (tatomyr) [inline packages/core/src/bundle/bundle-visitor.ts]: Why reassigning?

File: packages/core/src/bundle/bundle-visitor.ts [hunk 13]
```diff
@@ -291,22 +300,88 @@ export function makeBundleVisitor({
...
+    let name = prevName;
     let serialId = 2;
     while (componentsGroup[name] && !isEqualOrEqualRef(componentsGroup[name], target, ctx)) {
       name = `${prevName}-${serialId}`;
       serialId++;
     }
+    return { name, prevName };
+  }
+
+  function componentNameFromTitle(
```
Comment (tatomyr) [inline packages/core/src/bundle/bundle-visitor.ts:298]: Let's check if we can reuse the existing logic.

File: tests/e2e/bundle/bundle.test.ts [hunk 14]
```diff
@@ -200,3 +201,20 @@ describe('bundle with long description', () => {
...
+describe('bundle with option: use-titles-for-component-names', () => {
+  test('fails when two schemas share a title-based name', async () => {
+    const testPath = join(__dirname, 'bundle-use-titles-for-component-names-collision');
+    const entryPoints = getEntrypoints(testPath);
+    const args = [
+      indexEntryPoint,
+      'bundle',
+      '--use-titles-for-component-names',
+      '--component-renaming-conflicts-severity',
+      'error',
```
Comment (tatomyr) [inline tests/e2e/bundle/bundle.test.ts:215]: LEt's add `off` as well.

Comment (kanoru3101) [reply tests/e2e/bundle/bundle.test.ts:215]: Added test with `off`

File: packages/core/src/__tests__/utils.test.ts (hunk 11 above)
Comment (kanoru3101) [reply packages/core/src/__tests__/utils.test.ts]: The names `$input` and `$expected` have been changed for clarity

File: packages/core/src/__tests__/bundle.test.ts (hunk 10 above)
Comment (kanoru3101) [reply packages/core/src/__tests__/bundle.test.ts]: Changed to `toMatchInlineSnapshot` where checking error messages and left `toMatchSnapshot` for the happy path

File: packages/core/src/bundle/bundle-visitor.ts (hunk 12 above)
Comment (kanoru3101) [reply packages/core/src/bundle/bundle-visitor.ts]: Removed this reassigning

File: packages/core/src/bundle/bundle-visitor.ts (hunk 13 above)
Comment (kanoru3101) [reply packages/core/src/bundle/bundle-visitor.ts:298]: I tried collapsing it into a single if inside `getComponentName` without the extracted helper, but it comes out worse: the title path has failure modes the basename path doesn't (missing title / unusable chars / collision) and on failure it falls back to the basename, so you either duplicate the `-N` loop or scatter `useTitle && `checks through one big function. Keeping `componentNameFromBasename` as the shared primitive avoids both while leaving the default logic untouched.

File: packages/core/src/bundle/bundle-visitor.ts [hunk 15]
```diff
@@ -116,17 +122,23 @@ export function makeBundleVisitor({
...
   resolvedRefMap: ResolvedRefMap;
   keepUrlRefs: boolean;
   componentRenamingConflicts?: RuleSeverity;
+  useTitlesForComponentNames?: boolean;
 }) {
-  let components: Record<string, Record<string, unknown>>;
+  let components: Record<string, ComponentsGroup>;
   let rootLocation: Location;
 
+  const firstSchemaLocationByName = new Map<string, Location>();
```
Comment (kanoru3101) [inline packages/core/src/bundle/bundle-visitor.ts:139]: firstSchemaLocationByName exists only to populate the collision's from — the "referenced from …" link that points at the first schema which already took the name. It can't be derived the way the basename path handles a conflict, because the two collisions behave differently:
* Basename collision auto-resolves: it appends -N and reports a rename warning at the current location. It never references the other schema, so it needs no location tracking.
*  Title collision is a user-fixable error — the fix is to rename one of the two conflicting titles, so the message points from at the other (first) schema to help the user find both.

**Review by tatomyr — APPROVED** [review]
LGTM

**Comment by drwicid** [comment]
Question — Should lint also evaluate the effective derived component name?

My understanding is that bundled inlined schemas already receive derived component names today, traditionally based on the referenced file basename or $ref pointer segment. This option adds another derivation strategy for schemas: when --use-titles-for-component-names is enabled, the component name can instead come from the schema’s title, with fallback to the existing basename/pointer logic when the title is missing or invalid.

Given that the bundle process already has to resolve an effective component name either way, would it make sense for the lint/rule phase to expose or reuse that same derived-name calculation?

Linting can certainly validate the presence or shape of title fields directly. However, that only covers one input to the naming strategy. What seems more useful is linting against the final effective name that bundle would produce, regardless of whether that name came from title or from the traditional basename / $ref pointer fallback.

That would allow rules such as component-name-unique to detect collisions before bundling, including collisions introduced by title-derived names as well as collisions that still occur through fallback naming. It would also let validity and naming/casing rules apply to the actual component key that will be emitted, with diagnostics reported against the original schema files and source locations.

This seems preferable to requiring a second lint pass against the generated bundled file, where diagnostics are tied to generated output and are less directly traceable to the originating schema.

I understand the lifecycle split: bundle runs preprocessors, rules, and decorators, while lint runs preprocessors and rules but not decorators. Is keeping effective component-name resolution bundle-only an intentional boundary, or would sharing the derived-name helper with lint be reasonable so these checks can happen before bundling?

File: docs/@v2/commands/bundle.md [hunk 16]
```diff
@@ -45,6 +45,7 @@ redocly bundle --version
...
 | --skip-decorator                        | [string] | Ignore certain decorators. See the [Skip preprocessor, rule, or decorator section](#skip-preprocessor-rule-or-decorator).                                                                                                                                 |
 | --skip-preprocessor                     | [string] | Ignore certain preprocessors. See the [Skip preprocessor, rule, or decorator section](#skip-preprocessor-rule-or-decorator).                                                                                                                              |
+| --use-titles-for-component-names        | boolean  | Build Schema component names from each schema's `title` field. See [Use titles for component names](#use-titles-for-component-names).                                                                                                                     |
```
Comment (RomanHotsiy) [inline docs/@v2/commands/bundle.md]: Let's change it to --component-names-strategy with options like "basename" (default) or "title"?

@tatomyr what do you think?

File: packages/core/src/bundle/bundle-visitor.ts [hunk 17]
```diff
@@ -291,30 +294,97 @@ export function makeBundleVisitor({
...
+    const { node } = target;
+    const title = isPlainObject(node) && isString(node.title) ? node.title.trim() : '';
+    const key = toPascalCase(title);
+    const titleLocation = target.location.child('title');
+
+    if (title === '') {
+      return {
+        key,
+        problem: {
+          message: 'Schema must define a `title` to build a component name.',
```
Comment (RomanHotsiy) [inline packages/core/src/bundle/bundle-visitor.ts]: Need to clarify it is required only when using this option or someone may get confused:

```suggestion
          message: 'Schema must define a `title` when using --use-titles-for-component-names.',
```

File: packages/core/src/bundle/bundle-visitor.ts [hunk 18]
```diff
@@ -291,30 +294,97 @@ export function makeBundleVisitor({
...
+          location: target.location,
+          forceSeverity: 'error',
+        },
+      };
+    }
+    if (!new RegExp(COMPONENT_NAME_PATTERN).test(key)) {
```
Comment (RomanHotsiy) [inline packages/core/src/bundle/bundle-visitor.ts]: do we need it? should we just replace any unsupported symbol ourselves with dashes?

File: docs/@v2/commands/bundle.md (hunk 16 above)
Comment (tatomyr) [reply docs/@v2/commands/bundle.md]: Sounds even better to me. What about "filename" instead of "basename"?

File: packages/core/src/bundle/bundle-visitor.ts (hunk 18 above)
Comment (kanoru3101) [reply packages/core/src/bundle/bundle-visitor.ts]: I'll change it to replace and add to docs about it

**Comment by kanoru3101** [comment]
> Question — Should lint also evaluate the effective derived component name?
> 
> My understanding is that bundled inlined schemas already receive derived component names today, traditionally based on the referenced file basename or $ref pointer segment. This option adds another derivation strategy for schemas: when --use-titles-for-component-names is enabled, the component name can instead come from the schema’s title, with fallback to the existing basename/pointer logic when the title is missing or invalid.
> 
> Given that the bundle process already has to resolve an effective component name either way, would it make sense for the lint/rule phase to expose or reuse that same derived-name calculation?
> 
> Linting can certainly validate the presence or shape of title fields directly. However, that only covers one input to the naming strategy. What seems more useful is linting against the final effective name that bundle would produce, regardless of whether that name came from title or from the traditional basename / $ref pointer fallback.
> 
> That would allow rules such as component-name-unique to detect collisions before bundling, including collisions introduced by title-derived names as well as collisions that still occur through fallback naming. It would also let validity and naming/casing rules apply to the actual component key that will be emitted, with diagnostics reported against the original schema files and source locations.
> 
> This seems preferable to requiring a second lint pass against the generated bundled file, where diagnostics are tied to generated output and are less directly traceable to the originating schema.
> 
> I understand the lifecycle split: bundle runs preprocessors, rules, and decorators, while lint runs preprocessors and rules but not decorators. Is keeping effective component-name resolution bundle-only an intentional boundary, or would sharing the derived-name helper with lint be reasonable so these checks can happen before bundling?

Hi @drwicid 
Good question. If I read it right: you'd like lint to reuse bundle's effective-name calculation to catch collisions before bundling.

But before bundle the schemas live in separate files — there's no single combined namespace yet, so the real names (and their collisions) only appear once bundle merges everything. A pre-bundle rule like `component-name-unique` can only approximate them: it compares basename keys and ignores title. Matching bundle's exact names would mean re-running its stateful assembly (-N, title→fallback, dedup) in lint — a fair follow-up, but bigger than this flag.
 
 And collisions are caught at bundle anyway: bundle reports them, `--component-renaming-conflicts-severity=error` makes a collision fail the bundle, and a missing title already fails it — so nothing slips through.

File: docs/@v2/commands/bundle.md [hunk 19]
```diff
@@ -147,3 +148,59 @@ You can adjust how the CLI handles these naming conflicts with the `--component-
...
+```
+
+The output uses `OrderModel` and `OrderRequest`.
+
+{% /tab  %}
+{% /tabs  %}
+
+{% admonition type="warning" name="Title sanitization" %}
+The OpenAPI and AsyncAPI specifications allow only ASCII letters (`a`–`z`, `A`–`Z`), digits, `.`, `-`, and `_` in a Components Object key.
+Those characters are kept; any other character — including non-ASCII letters such as `é` or `я` — is replaced with `-` (for example, `User & Group` becomes `User-Group`).
```
Comment (JLekawa) [inline docs/@v2/commands/bundle.md]:
```suggestion
All other characters, including non-ASCII letters such as `é` or `я`, are replaced with `-` (for example, `User & Group` becomes `User-Group`).
```

File: .changeset/bright-banks-wish.md [hunk 20]
```diff
@@ -0,0 +1,7 @@
+---
+'@redocly/openapi-core': minor
+'@redocly/cli': minor
+---
+
+Added the `--component-names-strategy` option to the `bundle` command.
+THis option allows a choice of how inline Schema components are named: `basename` (default) or `title` (from each schema's `title` field).
```
Comment (JLekawa) [inline .changeset/bright-banks-wish.md]:
```suggestion
This option allows a choice of how inline Schema components are named: `basename` (default) or `title` (from each schema's `title` field).
```

File: docs/@v2/commands/bundle.md [hunk 21]
```diff
@@ -147,3 +148,59 @@ You can adjust how the CLI handles these naming conflicts with the `--component-
...
+A schema without a `title` can't be named this way, so bundling reports an error for that schema.
```
Comment (JLekawa) [inline docs/@v2/commands/bundle.md]:
```suggestion
Schemas without `title` can't be named using the `--component-names-strategy=title` strategy.
The bundling process reports an error for such schemas.
```

File: docs/@v2/commands/bundle.md (hunk 16 above)
Comment (RomanHotsiy) [reply docs/@v2/commands/bundle.md]: I'm fine with either

**Review by RomanHotsiy — APPROVED** [review]
